index.mjs

```diff
@@ -13,6 +14,24 @@ export {
   invertPatch
 };
 
+class TimeoutError extends Error {
+  constructor() {
+    super('timeout reached');
+  }
+}
+
+function assertTimeout(startTime, msTimeout) {
+  if (msTimeout !== undefined && Date.now() - startTime > msTimeout) {
+    throw new TimeoutError();
+  }
+}
+
+function calcRemainingTimeout(startTime, msTimeout) {
+  if (msTimeout === undefined) {
+    return undefined;
+  }
+  return msTimeout + startTime - Date.now();
+}
```

Comment:
I think this implementation is taking too many readings of `Date.now`, e.g - most places that call `calcRemainingTimeout` will also take another reading of `Date.now` immediately afterwards as it enters the next function.

maybe instead of taking the start time separately every time, support getting the timeout as a number or an assertion function.
if you got a number, create the assertion function (that would be the only case where you need to take the start time), if you got a function, you can just use it without having to calculate remaining time or taking a new start time.
e.g
```typescript
function getAssertTimeout(timeout) {
  if (typeof timeout === 'number') {
    const endTime = Date.now() + timeout
    return () => { if(Date.now() > endTime) throw new TimeoutError() }
  }
  if (typeof timeout === 'function') {
    return timeout
  }
  return () => {}
}
```
and then use that to get the assertion function in every place where you currently get a timeout and take a `startTime = Date.now()`